src/TokenManager.sol

```diff
@@ -0,0 +1,166 @@
+        uint128 amount
+    ) public {
+        LiquidityPoolLike liquidityPool =
+            LiquidityPoolLike(investmentManager.liquidityPools(poolId, trancheId, currency));
+        require(address(liquidityPool) != address(0), "TokenManager/unknown-token");
+
+        require(liquidityPool.balanceOf(msg.sender) >= amount, "TokenManager/insufficient-balance");
+        liquidityPool.burn(msg.sender, amount);
+
+        gateway.transferTrancheTokensToCentrifuge(poolId, trancheId, msg.sender, destinationAddress, amount);
+    }
+
+    function transferTrancheTokensToEVM(
+        uint64 poolId,
+        bytes16 trancheId,
+        uint64 destinationChainId,
+        address destinationAddress,
+        uint128 amount
+    ) public {
+        // TODO: this should get the underlying tranche token once the single tranche token change has been merged
+        LiquidityPoolLike liquidityPool =
+            LiquidityPoolLike(investmentManager.getLiquidityPoolsForTranche(poolId, trancheId)[0]);
+        require(address(liquidityPool) != address(0), "TokenManager/unknown-token");
+
+        require(liquidityPool.balanceOf(msg.sender) >= amount, "TokenManager/insufficient-balance");
+        liquidityPool.burn(msg.sender, amount);
+
+        gateway.transferTrancheTokensToEVM(
+            poolId, trancheId, msg.sender, destinationChainId, destinationAddress, amount
+        );
+    }
+
+    // --- Incoming message handling ---
+    /// @dev a global chain agnostic currency index is maintained on centrifuge chain. This function maps a currency from the centrifuge chain index to its corresponding address on the evm chain.
+    /// The chain agnostic currency id has to be used to pass currency information to the centrifuge chain.
+    /// @notice this function can only be executed by the gateway contract.
+    function addCurrency(uint128 currency, address currencyAddress) public onlyGateway {
+        // currency index on the centrifuge chain side should start at 1
+        require(currency > 0, "TokenManager/currency-id-has-to-be-greater-than-0");
+        require(currencyIdToAddress[currency] == address(0), "TokenManager/currency-id-in-use");
+        require(currencyAddressToId[currencyAddress] == 0, "TokenManager/currency-address-in-use");
+
+        currencyIdToAddress[currency] = currencyAddress;
+        currencyAddressToId[currencyAddress] = currency;
+
+        // enable connectors to take the currency out of escrow in case of redemptions
+        EscrowLike(escrow).approve(currencyAddress, address(investmentManager), type(uint256).max);
+        emit CurrencyAdded(currency, currencyAddress);
+    }
+
+    function handleTransfer(uint128 currency, address recipient, uint128 amount) public onlyGateway {
+        address currencyAddress = currencyIdToAddress[currency];
+        require(currencyAddress != address(0), "TokenManager/unknown-currency");
+
+        EscrowLike(escrow).approve(currencyAddress, address(this), amount);
```

Comment:
I would add the full approval within the `addCurrency` function when adding a new currency. Same as we are doing for the investmentManager

`EscrowLike(escrow).approve(currencyAddress, address(this), type(uint256).max);`